enhanced/scraper.go

```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
-			l := s.logger.With("EventId", *event.EventId).With("LogStreamName", *event.LogStreamName)
-			l = l.With("Timestamp", aws.MillisecondsTimeValue(event.Timestamp).UTC())
-			l = l.With("IngestionTime", aws.MillisecondsTimeValue(event.IngestionTime).UTC())
-
-			var instance *sessions.Instance
-			for _, i := range s.instances {
-				if i.ResourceID == *event.LogStreamName {
-					instance = &i
-					break
+
+	// LogStreamNames parameter supports up to 100 items.
+	// https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html
+	for i := 0; len(s.logStreamNames)-(100*i) > 0; i++ {
+		slice_start := i * 100
+		slice_end := slice_start + 100
+		if slice_end > len(s.logStreamNames) {
+			slice_end = len(s.logStreamNames)
+		}
+
+		input := &cloudwatchlogs.FilterLogEventsInput{
+			LogGroupName:   aws.String("RDSOSMetrics"),
+			LogStreamNames: aws.StringSlice(s.logStreamNames[slice_start:slice_end]),
+			StartTime:      aws.Int64(aws.TimeUnixMilli(s.nextStartTime)),
+		}
+		s.logger.Debugf("Requesting metrics since %s (last %s).", s.nextStartTime.UTC(), time.Since(s.nextStartTime))
+
+		collectAllMetrics := func(output *cloudwatchlogs.FilterLogEventsOutput, lastPage bool) bool {
+			for _, event := range output.Events {
+				l := s.logger.With("EventId", *event.EventId).With("LogStreamName", *event.LogStreamName)
+				l = l.With("Timestamp", aws.MillisecondsTimeValue(event.Timestamp).UTC())
+				l = l.With("IngestionTime", aws.MillisecondsTimeValue(event.IngestionTime).UTC())
+
+				var instance *sessions.Instance
+				for _, i := range s.instances {
+					if i.ResourceID == *event.LogStreamName {
+						instance = &i
+						break
+					}
 				}
-			}
-			if instance == nil {
-				l.Errorf("Failed to find instance.")
-				continue
-			}
-			l = l.With("region", instance.Region).With("instance", instance.Instance)
+				if instance == nil {
+					l.Errorf("Failed to find instance.")
+					continue
+				}
+				l = l.With("region", instance.Region).With("instance", instance.Instance)
 
-			// l.Debugf("Message:\n%s", *event.Message)
-			osMetrics, err := parseOSMetrics([]byte(*event.Message))
-			if err != nil {
-				l.Errorf("Failed to parse metrics: %s.", err)
-				continue
-			}
-			// l.Debugf("OS Metrics:\n%#v", osMetrics)
+				// l.Debugf("Message:\n%s", *event.Message)
+				osMetrics, err := parseOSMetrics([]byte(*event.Message))
+				if err != nil {
+					l.Errorf("Failed to parse metrics: %s.", err)
+					continue
+				}
+				// l.Debugf("OS Metrics:\n%#v", osMetrics)
 
-			if allMetrics[instance.ResourceID] == nil {
-				allMetrics[instance.ResourceID] = make(map[time.Time][]prometheus.Metric)
+				if allMetrics[instance.ResourceID] == nil {
+					allMetrics[instance.ResourceID] = make(map[time.Time][]prometheus.Metric)
+				}
+				timestamp := aws.MillisecondsTimeValue(event.Timestamp)
+				metrics := osMetrics.makePrometheusMetrics(instance.Region)
+				allMetrics[instance.ResourceID][timestamp] = metrics
```

Comment:
About the “do not paralelize this code”, in particular all of this code would all need to be done holding a mutex lock.

It is way better to just do things synchronously, until and unless you have profiling information showing that running this synchronously is performance bottleneck.